valhalla/userrequests/state_changes.py

```diff
@@ -141,6 +145,93 @@ def modify_ipp_time_from_requests(ipp_val, requests_list, modification='debit'):
         logger.warn(_("Problem {}ing ipp time for request {}: {}").format(modification, request.id, repr(e)))
 
 
+def get_request_state_from_pond_blocks(request_state, request_blocks):
+    active_blocks = False
+    future_blocks = False
+    now = timezone.now()
+    for block in request_blocks:
+        start_time = dateutil.parser.parse(block['start']).replace(tzinfo=timezone.utc)
+        end_time = dateutil.parser.parse(block['end']).replace(tzinfo=timezone.utc)
+        if all([molecule['complete'] for molecule in block['molecules']]):
+            return 'COMPLETED'
+        if (not block['canceled'] and not any([molecule['failed'] for molecule in block['molecules']])
+            and start_time < now < end_time):
+            active_blocks = True
+        if now < start_time:
+            future_blocks = True
+
+    if not (future_blocks or active_blocks):
+        return 'FAILED'
+
+    return request_state
+
+
+def get_request_state(request_state, request_blocks, ur_expired):
+    if request_state == 'COMPLETED':
+        return request_state
+    else:
+        pond_state = get_request_state_from_pond_blocks(request_state, request_blocks)
+        if pond_state in ['COMPLETED', 'WINDOW_EXPIRED', 'CANCELED']:
```

Comment:
Either we have this check on this line, or we have it on line 177. It is needed because ur_expired could be true, but we don't want to set the state as 'WINDOW_EXPIRED' if it should be 'COMPLETED' or 'CANCELED'